### PR TITLE
JAMES-2529 [FILTER] Mailet integration tests

### PR DESCRIPTION
**Pull request by chibenwa**
Ships a minimal test subset at the integration level. Edge cases are handled in detail in unit tests in https://github.com/linagora/james-project/pull/1683

Note that I also:

 - Made the mailet mandatory
 - Configured it in the various places
 - And documented it

### CODE REVIEW

**Comment by chibenwa** [comment]
```
DefaultCassandraJamesServerTest.memoryJamesServerShouldStartWithNoConfigurationFile:60 » Runtime

[19f6fe1e2da07d4ae8a9382bd9a978dbb002cd49] [INFO] Apache James :: Server :: Cassandra - guice injection FAILURE [02:14 min]
```

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/JmapCommonRequests.java [hunk 1]
```diff
@@ -85,6 +86,24 @@ public static boolean isAnyMessageFoundInRecipientsMailboxes(AccessToken recipie
         }
     }
 
+    public static boolean isAnyMessageFoundInRecipientsMailbox(AccessToken recipientToken, MailboxId mailboxId) {
+        try {
+            with()
+                .header("Authorization", recipientToken.serialize())
+                    .body("[[\"getMessageList\", {\"filter\":{\"inMailboxes\":[\"" + mailboxId.serialize() + "\"]}}, \"#0\"]]")
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/JmapCommonRequests.java]: indent

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 2]
```diff
@@ -572,4 +591,1159 @@ public void allFieldsAndComparatorsShouldBeSupported() {
...
+                "      \"action\": {" +
+                "        \"appendIn\": {" +
+                "          \"mailboxIds\": [\"" + matchedMailbox.serialize() + "\"]" +
+                "        }" +
+                "      }" +
+                "    }" +
+                "  ]}, " +
+                "\"#0\"" +
+                "]]")
+            .post("/jmap");
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: You may assert at least the status code of this response.
If this request fails then it will be hard to understand why the assert of this test is failing.

File: server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPModule.java [hunk 3]
```diff
@@ -96,7 +101,8 @@ protected void configure() {
 
         Multibinder<CamelMailetContainerModule.TransportProcessorCheck> transportProcessorChecks = Multibinder.newSetBinder(binder(), CamelMailetContainerModule.TransportProcessorCheck.class);
         transportProcessorChecks.addBinding().toInstance(VACATION_MAILET_CHECK);
-        
+        transportProcessorChecks.addBinding().toInstance(FILTERING_MAILET_CHECK);
```
Comment (aduprat) [inline server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPModule.java]: Then we will break the production ;)

Comment (chibenwa) [reply server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPModule.java]: No problem :trollface:

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 4]
```diff
@@ -572,4 +591,1159 @@ public void allFieldsAndComparatorsShouldBeSupported() {
...
+
+        with()
+            .header("Authorization", bobAccessToken.serialize())
+            .body(requestBody)
+            .post("/jmap");
+
+        calmlyAwait.until(
+            () -> JmapCommonRequests.isAnyMessageFoundInRecipientsMailbox(accessToken, inbox));
+    }
+
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: you are always testing that if the rule matches, it goes to folder. If it does not, it goes to inbox.
Can you add some tests:
- if it matches it goes to folder, but not in inbox
- if it does not match, it goes to inbox, but not in folder

Can you also add at least one test with several filters?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 5]
```diff
@@ -616,7 +616,9 @@ public void messageShouldBeAppendedInSpecificMailboxWhenFromRuleMatches() {
                 "  ]}, " +
                 "\"#0\"" +
                 "]]")
-            .post("/jmap");
+            .post("/jmap")
+            .then()
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: indent

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 6]
```diff
@@ -572,4 +591,1249 @@ public void allFieldsAndComparatorsShouldBeSupported() {
             .body(ARGUMENTS + ".singleton[4].condition.comparator", equalTo("contains"));
     }
 
+    @Test
+    public void messageShouldBeAppendedInSpecificMailboxWhenFromRuleMatches() {
+        with()
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: given()?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 7]
```diff
@@ -572,4 +591,1249 @@ public void allFieldsAndComparatorsShouldBeSupported() {
...
+                "      \"action\": {" +
+                "        \"appendIn\": {" +
+                "          \"mailboxIds\": [\"" + matchedMailbox.serialize() + "\"]" +
+                "        }" +
+                "      }" +
+                "    }" +
+                "  ]}, " +
+                "\"#0\"" +
+                "]]")
+            .post("/jmap")
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: missing when()?

**Comment by chibenwa** [comment]
```
[3d4cab6df2b996dc44f0222ef44de79729e84c90] [INFO] Apache James :: Server :: Cli :: Integretion ....... FAILURE [ 40.968 s]

[3d4cab6df2b996dc44f0222ef44de79729e84c90] [ERROR] setSieveUserQuotaShouldWork(org.apache.james.cli.SieveQuotaCommandsIntegrationTest)  Time elapsed: 0.338 s  <<< ERROR!
[3d4cab6df2b996dc44f0222ef44de79729e84c90] java.lang.RuntimeException: org.apache.commons.configuration.ConfigurationException: Missing org.apache.james.jmap.mailet.filter.JMAPFiltering in mailets configuration (mailetcontainer -> processors -> transport). 
[3d4cab6df2b996dc44f0222ef44de79729e84c90] 	at org.apache.james.cli.SieveQuotaCommandsIntegrationTest.setUp(SieveQuotaCommandsIntegrationTest.java:56)
```

**Comment by aduprat** [comment]
merged
